src/main/java/org/springframework/data/mapping/model/KotlinValueUtils.java

```diff
@@ -0,0 +1,397 @@
+		return new ValueBoxing(BoxingRules.CONSTRUCTOR, parameter);
+	}
+
+	/**
+	 * Creates a value hierarchy across value types from a given {@link KParameter} for constructor usage.
+	 *
+	 * @param cls the entrypoint of the type hierarchy.
+	 * @return the {@link ValueBoxing} type hierarchy.
+	 */
+	public static ValueBoxing getConstructorValueHierarchy(Class<?> cls) {
+
+		KClass<?> kotlinClass = JvmClassMappingKt.getKotlinClass(cls);
+		return new ValueBoxing(BoxingRules.CONSTRUCTOR, Reflection.typeOf(kotlinClass), kotlinClass, false);
+	}
+
+	/**
+	 * Boxing rules for value class wrappers.
+	 */
+	enum BoxingRules {
+
+		/**
+		 * When used in the constructor. Constructor boxing depends on nullability of the declared property, whether the
+		 * component uses defaulting, nullability of the value component and whether the component is a primitive.
+		 */
+		CONSTRUCTOR {
+			@Override
+			public boolean shouldApplyBoxing(KType type, boolean optional, KParameter component) {
+
+				Type javaType = ReflectJvmMapping.getJavaType(component.getType());
+				boolean isPrimitive = javaType instanceof Class<?> c && c.isPrimitive();
+
+				if (type.isMarkedNullable() || optional) {
+					return (isPrimitive && type.isMarkedNullable()) || component.getType().isMarkedNullable();
+				}
+
+				return false;
+			}
+		},
+
+		/**
+		 * When used in a copy method. Copy method boxing depends on nullability of the declared property, nullability of
+		 * the value component and whether the component is a primitive.
+		 */
+		COPY {
+			@Override
+			public boolean shouldApplyBoxing(KType type, boolean optional, KParameter component) {
+
+				KType copyType = expandUnderlyingType(type);
+
+				if (copyType.getClassifier()instanceof KClass<?> kc && kc.isValue() || copyType.isMarkedNullable()) {
+					return true;
+				}
+
+				return false;
+			}
+
+			private static KType expandUnderlyingType(KType kotlinType) {
+
+				if (!(kotlinType.getClassifier()instanceof KClass<?> kc) || !kc.isValue()) {
+					return kotlinType;
+				}
+
+				List<KProperty<?>> properties = getProperties(kc);
+				if (properties.isEmpty()) {
+					return kotlinType;
+				}
+
+				KType underlyingType = properties.get(0).getReturnType();
+				KType componentType = ValueBoxing.resolveType(underlyingType);
+				KType expandedUnderlyingType = expandUnderlyingType(componentType);
+
+				if (!kotlinType.isMarkedNullable()) {
+					return expandedUnderlyingType;
+				}
+
+				if (expandedUnderlyingType.isMarkedNullable()) {
+					return kotlinType;
+				}
+
+				Type javaType = ReflectJvmMapping.getJavaType(expandedUnderlyingType);
+				boolean isPrimitive = javaType instanceof Class<?> c && c.isPrimitive();
+
+				if (isPrimitive) {
+					return kotlinType;
+				}
+
+				return expandedUnderlyingType;
+			}
+
+			static List<KProperty<?>> getProperties(KClass<?> kClass) {
+
+				if (kClass.isValue()) {
+
+					for (KCallable<?> member : kClass.getMembers()) {
```

Comment:
@mp911de Similarly to the above, this should use `memberProperties`, otherwise it'll fail on something like this:
```
@JvmInline
value class Z(val value: String) {
    val String.unrelated: String get() = this
}
```